Multi selection - implements #22

geowarin opened the pull request:
Implements #22.

I've added tests for every use case detailed in the issue.
There are a couple of things that can be improved immediately :

* Since we have the **element()** method, I think we can safely remove the params methods in tests.js which becomes somewhat obsolete. This will allow to remove the ugly **unarray** method
* The find middleware has become a bit more complicated but I think that's ok. Anyway, we can find much better names for the functions declared in there

I'm looking forward for your reviews and comments :)

geowarin commented:
I also added two middlewares:

## The Index middleware

Allows to select a specific element after a multi find

    Test(<TableComponent/>)
      .find('table', {multi: false})
      .find('td', {multi: true})
      .index(1)
      .element(td => {
        expect(td.getDOMNode().textContent).to.equal('2')
      })

## The setValues middleware

    Test(<FormComponent />)
      .find('input', {multi: true})
      .setValues(['username', 'password'])

zackify commented:
I'll look this over tonight!

zackify commented on src/tests.js:
```diff
@@ -3,7 +3,19 @@ let { TestUtils } = React.addons
-import {Find, SetState, Simulate} from './middleware'
+import {Find, SetState, Simulate, SetValues, Index} from './middleware'
+
+const unarray = (elems) => {
```
Since we're making breaking changes, might as well remove this like you said.

zackify commented:
So it all looks good to me, just remove the params method, update the readme, add a list of the breaking changes, and I'll merge it unless if @dphaener or @kenwheeler see anything wrong. We need to start a wiki with all of the options and examples

vramana commented on src/middleware/find.js:
```diff
@@ -1,18 +1,62 @@
-import React from 'react'
-let { TestUtils } = React.addons
+import React from 'react';
```
Why add semicolons here?? May be you can add no-semicolon rule to the eslintrc. So, the style is consistent across the codebase.

zackify replied:
Agreed, call me evil, but let's remove them all.

vramana commented on src/tests.js:
```diff
@@ -65,18 +77,28 @@ export default class Test {
-  find(data){
-    Find.call(this, data)
+  find(){
+    Find.call(this, ...arguments)
```
You should write this as
```javascript 
find(...args) {
   Find.call(this, ...args)
   return this
}
```

vramana commented on tests/components/button-component.jsx:
```diff
@@ -0,0 +1,19 @@
+  render() {
+    return <div>
+      <div className="main">
+	<button value="ok" onClick={this.props.action1} />
```
I think you should indent here.

vramana commented on tests/components/component.jsx:
```diff
@@ -12,14 +12,14 @@ export default class TestComponent extends React.Component {
-        type="button"
-        onClick={this.props.onClick}>
-        Click Me
+	type="button"
```
Dont remove the indentation.

zackify commented on .editorconfig:
```diff
@@ -0,0 +1,9 @@
+root = true
```
add this to .npmignore

vramana commented on tests/components/form-component.jsx:
```diff
@@ -0,0 +1,28 @@
+export default class TestComponent extends Component {
+  constructor() {
+    super()
+    this.state = {}
```
```javascript
constructor(...args) {
  super(...args)
  this.state = {}
}
```

or 

```javascript
constructor(props, context) {
  super(props, context)
  this.state = {}
}
```
Similarly at other places.

zackify replied:
this doesn't really matter that much, it's just for testing

vramana replied:
We don't lose anything by maintaining consistency I feel.

zackify replied:
agreed :)

vramana commented on tests/components/form-component.jsx:
```diff
@@ -0,0 +1,28 @@
+    );
+  }
+
+  handleSubmit() {
```
Keep the `render` method at the bottom.

zackify replied:
:+1:

vramana commented on tests/index.jsx:
```diff
@@ -0,0 +1,25 @@
+
+  it('should throw an error if the root is not multi', () => {
+
+    var search = () => Test(<TableComponent/>).find('table', {multi: false}).index(0)
```
Use `const`

vramana replied:
@geowarin I think you missed this. Also Change the file name from `index.jsx` to something else. It is slightly misleading at first site becaues index.js is used to export stuff but this concers with index middleware.

zackify commented on src/middleware/find.js:
```diff
@@ -1,18 +1,62 @@
 
-export default function find(selector){
-  let elements, name
+function doFind(root, selector) {
```
let's call this instance, since it is the component instance?

vramana commented on src/middleware/find.js:
```diff
@@ -1,18 +1,62 @@
+
+function findSingleOrMulti(root, selector) {
+  if (Array.isArray(root)) {
+    var matches = root.map(element => doFind(element, selector));
```
Use `const`. Also prefer `const` over `let`.

vramana commented on src/middleware/find.js:
```diff
@@ -1,18 +1,62 @@
+
+export default function find(selector, options) {
+  let fromRoot = options && options.root;
+  let multi = options && options.multi;
```
You can write this as const { multi } = options

vramana commented on src/tests.js:
```diff
@@ -52,7 +63,8 @@ export default class Test {
     if(this.helpers.elements && length === 1) {
-      return Object.assign({}, this, this.helpers.elements)
+      let elements = unarray(this.helpers.elements)
+      return Object.assign({}, this, elements)
```
Can be written as `{...this, ...elements}`

vramana commented:
Excepts for storing `elements` I think `helpers` is never used. Why dont we just have simply this.elements instead of `this.helpers.elements` ? That extra layer of indirection is complicates the code I feel.

geowarin commented:
Hi guys!
Thank you very much for your inputs!

I believe I've addressed the concerns you gentlemen had with the coding style.
I also added rules to enforce the "no-semi" rule and the "prefer const" rule so nobody can have the indelicacy of messing up with the coding style like I did :smile: 

I also updated the README presuming that the next release would be 0.5.

Before you merge, please make sure that you won't regret pushing the new syntax. My good friend and colleague @cedbossneo suggested the following which removes the needs for the options in the `find()` middleware :

    Test(<FormComponent login={spy}/>)
      .find('input') // find is always multi by default
      .setValues(['username', 'password'])
      .root() // replaces the {root: true} option
      .find('form')
      .first() // or last() or index()
      .simulate('submit')
      .test(function () {
        expect(spy.calledWith('username', 'password')).to.be.true;
      })

Any thoughts?

geowarin commented:
@vramana Agreed.

vramana commented on src/tests.js:
```diff
@@ -29,54 +30,53 @@ export default class Test {
 
-  find(data){
-    Find.call(this, data)
+  find(...args){
```
Why `...args` ?? We are only supposed pass a single variable here. I think it should `find(selector)`.

vramana commented on tests/components/component.jsx:
```diff
@@ -0,0 +1,27 @@
+
+  render() {
+    return <body>
+    <div onClick={this.props.onClick}>{this.state.test}</div>
```
Indent properly here. Use `div` instead of `body`

geowarin replied:
This changes mysteriously breaks some tests unrelated to this pull request.
Maybe @zackify can take a look

vramana commented on tests/components/table-component.jsx:
```diff
@@ -0,0 +1,21 @@
+import React from 'react'
+
+export default class TableComponent extends React.Component {
```
Import `Component` also at the top level.

vramana commented on src/middleware/find.js:
```diff
@@ -1,18 +1,61 @@
+    throw new Error(`Could not find element "${selector}"`)
+
+  if (!this.helpers.elements) {
+    this.helpers.elements = []
```
Why initialise it to an array?? If you are using it as an object below ?? Also This should be part of constructor not here.

vramana commented on src/middleware/find.js:
```diff
@@ -1,18 +1,61 @@
+  const {elements, name} = findSingleOrMulti(rootElement, selector)
+
+  if (!elements || elements.length === 0)
+    throw new Error(`Could not find element "${selector}"`)
```
I think it is necessary to throw. Because the element[selector] will be undefined That is good enough check. Is there more rationale in an throwing error??

geowarin replied:
The rationale is simply to have more descriptive errors

vramana replied:
Isn't it 'Could not find element selector' same as being `elements[selector]` being `undefined` ?? I don't see what extra information it carries.

geowarin replied:
I kind of agree but this requires more intimate knowledge of the framework.
Moreover, throwing an exception allows writing a test to assert this behaviour.
Let's ask @zackify :smiley:

vramana commented on src/middleware/find.js:
```diff
@@ -1,18 +1,61 @@
+  }
+}
+
+const flatten = array => {
```
`const flatten = array => [].concat.apply([], array)`

vramana commented on tests/components/component.jsx:
```diff
@@ -1,13 +1,13 @@
-import React from 'react'
+import {Component} from 'react'
```
You need to do import React, { Component } from 'react'

geowarin replied:
Why? The React object is not used in those components

vramana replied:
You are. All the JSX components you write are converted to `React.createElement('div', ...)` Babel transpiles this but it doesn't import it.

geowarin commented on src/middleware/find.js:
```diff
@@ -1,18 +1,56 @@
+  const {elements, name} = findSingleOrMulti(rootElement, selector)
+
+  if (elements.length === 0)
+    throw new Error(`Could not find element "${selector}"`)
```
So there was a debate about this throw between me and @vramana.
I think we should throw early to improve user feedback and be able to write a test to assert this behavior.

On the other hand if we don't throw, elements[name] will be undefined somewhere later.
@zackify, thought?

zackify replied:
Throwing an error there is fine with me

geowarin commented on tests/components/component.jsx:
```diff
@@ -1,13 +1,13 @@
 
-  render(){
+  render() {
     return <body>
```
If changed to a div, this will break tests in find.jsx. Since this change would be somewhat unrelated to this pull request, maybe @zackify should take a look a this :smile:

zackify replied:
Body should be changed to div. we shouldn't render the body in react. If
you look at the react 14 branch, it's changed there because it caused a lot
of failing tests.
On Sat, Sep 26, 2015 at 02:59 Geoffroy Warin <notifications@github.com>
wrote:

> In tests/components/component.jsx
> <https://github.com/Legitcode/tests/pull/23#discussion_r40492257>:
>
> >      this.state = {}
> >    }
> >
> > -  render(){
> > +  render() {
> >      return <body>
>
> If changed to a div, this will break tests in find.jsx. Since this change
> would be somewhat unrelated to this pull request, maybe @zackify
> <https://github.com/zackify> should take a look a this [image: :smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Legitcode/tests/pull/23/files#r40492257>.
>

zackify commented on README.md:
```diff
@@ -94,4 +105,85 @@ Test(<TestComponent/>)
+  })
+~~~
+
+But if you use the `multi` option, you can find nested elements:
```
You should make multi the default, like you said. Also, I don't think I like the root option. The component instance is always available inside the test method if you need it

vramana commented on test/middlewares/find.jsx:
```diff
@@ -0,0 +1,45 @@
+    Test(<TestComponent/>)
+    .find('div')
+    .test(function() {
+      expect(this.elements.div.props.children).to.be.equal(undefined)
```
@zackify Is `test` method even necessary?? The code seems a bit convoluted. `element` should be prefered of writing tests. Choose the elements you want to test in the elements args and write tests on them.

geowarin commented:
Ok guys, the final API is like this:

~~~js
Test(<FormComponent login={spy}/>)
  .find('input').setValues(['username', 'password']) // multi by default
  .and() // come back to the root
  .find('form').first().simulate('submit')
  .test(function () {
    expect(spy.calledWith('username', 'password')).to.be.true;
  })
~~~

You can now use the `first()`, `last()` and `index()` middleware if you want to work on a specific element.

What do you think?

vramana commented:
In your example, lets say the sumbit returns error and we want to test with new values for username and password. Then I have to do `find('input')`, `find('form')`, `simulate('submit')` all over again. But we already have references for these elements, Why do we have to find them? 

One way to avoid this is to return is to check if the element is present in elements. But this doesn't allow `form` element to be over overridden by some other form inside the component. I feel that we are testing a component witlh lots of elements it is slightly hard to this via this api.

For finding nested elements, isn't it simpler to write a query like this `find('table td')` Then the immediate question is what will be the name of it in `elements`. May be we can pass an extra argument specifying name of the element i.e, `find('cells', 'table td')`. 

Are there any generic cases where `test` method is useful ?? We can do most of things by `element` method. I feel it should be the way to go because having one unambiguous way to write tests helps.

vramana commented on src/middleware/and.js:
```diff
@@ -0,0 +1,3 @@
+export default function And() {
+  this.root = this.instance
```
This should be a method inside Test class because it doesn't fit the definition of a middleware.

geowarin commented:
@vramana I like your suggestions, especially the possibility to name the elements that you find.
However, this does not allow naming only the first button for instance if I have a bunch of them.

Right now, the test() method is useful to assert things on the component like its props or state.
But we could create an instance() method that injects the component instance.

geowarin commented:
I also fear that leaving the possibility for the user to find something complicated like find('table .nested td') will make them want to do find('table > td.nested') and we will soon be reinventing jquery selectors.

vramana commented:
@geowarin I don't think we need to reinvent them. Even I was thinking this for a while. I think I found a solution. Once we have the DOM for `table` we can use querySelectorAll on it to filter of elements. I think it will work in most cases. May be I am missing something. I'll think more about this.

geowarin commented:
That would be awesome, except you cannot call ReactUtils.simulate() on a DOM node, can you?

vramana commented:
Yeah. This is going to be hard I guess. Let me think what can be done.

geowarin commented:
This has become a messy PR. I'm closing this waiting for #24 to rebase upon.
I invite @vramana, @zackify and any person who is interested in proposing an API for nested searches to continue the discussion here: #22.

Of course, I am volunteering for a new PR :smile:
